pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+			}
+			_, err = a.forwardingRulesClient.SetTarget(ctx, setTargetReq)
+		}
+		if err != nil {
+			return fmt.Errorf("updating ComputeForwardingRule target %s failed: %w", a.resourceID, err)
+		}
+		log.V(2).Info("successfully updated ComputeForwardingRule target", "name", a.resourceID)
+	}
+
+	updated := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		getReq := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		updated, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+	} else {
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		updated, err = a.forwardingRulesClient.Get(ctx, getReq)
+	}
+	if err != nil {
+		if direct.IsNotFound(err) {
+			return err
+		}
+		return fmt.Errorf("getting ComputeForwardingRule %q failed: %w", a.resourceID, err)
+	}
+
+	status := &krm.ComputeForwardingRuleStatus{}
+	// TODO(yuhou): Do we need other status beside conditions and observedGeneration?
+	status.SelfLink = updated.SelfLink
+	return setStatus(u, status)
+}
+
+func (a *forwardingRuleAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	// TODO(kcc)
+	return nil, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *forwardingRuleAdapter) Delete(ctx context.Context) (bool, error) {
+	if a.resourceID == "" {
+		return false, fmt.Errorf("resourceID is empty")
+	}
+
+	log := klog.FromContext(ctx).WithName(ctrlName)
+	log.V(2).Info("deleting ComputeForwardingRule", "name", a.resourceID)
+
+	exist, err := a.Find(ctx)
+	if err != nil {
+		return false, err
+	}
+	if !exist {
+		// return (false, nil) if the object was not found but should be presumed deleted.
+		return false, nil
+	}
+
+	if a.Location == "global" {
+		req := &computepb.DeleteGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		_, err = a.globalForwardingRulesClient.Delete(ctx, req)
+	} else {
+		req := &computepb.DeleteForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		_, err = a.forwardingRulesClient.Delete(ctx, req)
+	}
+	if err != nil {
+		return false, fmt.Errorf("deleting ComputeForwardingRule %s failed: %w", a.resourceID, err)
+	}
+	log.V(2).Info("successfully deleted ComputeForwardingRule", "name", a.resourceID)
+	return true, nil
+}
+
+func (a *forwardingRuleAdapter) globalFullyQualifiedName() string {
+	return fmt.Sprintf("projects/%s/global/forwardingRules/%s", a.projectID, a.resourceID)
+}
+
+func (a *forwardingRuleAdapter) fullyQualifiedName() string {
+	return fmt.Sprintf("projects/%s/regions/%s/forwardingRules/%s", a.projectID, a.Location, a.resourceID)
+}
+
+func (a *forwardingRuleAdapter) getParent() string {
+	// TODO(user): Write the GCP URI parent for your resource
+	return fmt.Sprintf("projects/%s", a.projectID)
```

Comment:
I think this should include global or the region.  It's also possible this isn't used though..

Reply:
I'm removing this method as it isn't used anywhere